django/contrib/admin/helpers.py

```diff
@@ -224,38 +224,44 @@ class InlineAdminFormSet(object):
     A wrapper around an inline formset for use in the admin system.
     """
     def __init__(self, inline, formset, fieldsets, prepopulated_fields=None,
-            readonly_fields=None, model_admin=None):
+            readonly_fields=None, uneditable_fields=None, has_add_permission=True,
+            model_admin=None):
         self.opts = inline
         self.formset = formset
         self.fieldsets = fieldsets
         self.model_admin = model_admin
         if readonly_fields is None:
             readonly_fields = ()
         self.readonly_fields = readonly_fields
+        if uneditable_fields is None:
+            uneditable_fields = ()
+        self.uneditable_fields = uneditable_fields
         if prepopulated_fields is None:
             prepopulated_fields = {}
         self.prepopulated_fields = prepopulated_fields
+        self.has_add_permission = has_add_permission
 
     def __iter__(self):
         for form, original in zip(self.formset.initial_forms, self.formset.get_queryset()):
             view_on_site_url = self.opts.get_view_on_site_url(original)
             yield InlineAdminForm(self.formset, form, self.fieldsets,
-                self.prepopulated_fields, original, self.readonly_fields,
+                self.prepopulated_fields, original, self.uneditable_fields,
                 model_admin=self.opts, view_on_site_url=view_on_site_url)
         for form in self.formset.extra_forms:
             yield InlineAdminForm(self.formset, form, self.fieldsets,
                 self.prepopulated_fields, None, self.readonly_fields,
                 model_admin=self.opts)
-        yield InlineAdminForm(self.formset, self.formset.empty_form,
-            self.fieldsets, self.prepopulated_fields, None,
-            self.readonly_fields, model_admin=self.opts)
+        if self.has_add_permission:
```

Comment:
I suppose, this condition is there because I had to change the behaviour of empty fields on the end of Inline formset. I You don't have add permissions and only view permissions You shouldn't be able to add anything. Before it was easy, because without change permissions, the inlines was not displayed at all, but with view permissions You have to display inlines in read only way.

BTW, can you please manage to add me to be one of the authors of the read-permissions commit?